File: frontend/src/api/warehouse.ts
import request from './request'
import type { PageResult } from './course'

export interface Warehouse {
  id: number
  name: string
  description: string | null
  warehouse_kind: 'file_type' | 'course'
  course_subject: string | null
  material_type: string
  icon: string
  color: string
  sort_order: number
  material_count: number
  created_at: string
  updated_at: string
}

export interface WarehouseMaterial {
  id: number
  course_id: number
  course_name: string
  warehouse_id: number | null
  type: string
  original_name: string
  status: string
  uploaded_by: number | null
  uploader_name: string
  created_at: string
}

export interface WarehouseCreate {
  name: string
  description?: string
  warehouse_kind?: 'file_type' | 'course'
  course_subject?: string
  material_type?: string
  icon?: string
  color?: string
  sort_order?: number
}

export interface WarehouseMaterialSearch {
  page_num?: number
  page_size?: number
  material_name?: string
  course_name?: string
  teacher_name?: string
  created_from?: string
  created_to?: string
}

export function listWarehouses(): Promise<Warehouse[]> {
  return request.get('/api/v1/warehouses')
}

export function getWarehouse(id: number): Promise<Warehouse> {
  return request.get(`/api/v1/warehouses/${id}`)
}

export function createWarehouse(data: WarehouseCreate): Promise<Warehouse> {
  return request.post('/api/v1/warehouses', data)
}

export function updateWarehouse(id: number, data: Partial<WarehouseCreate>): Promise<Warehouse> {
  return request.put(`/api/v1/warehouses/${id}`, data)
}

export function deleteWarehouse(id: number): Promise<null> {
  return request.delete(`/api/v1/warehouses/${id}`)
}

export function listWarehouseMaterials(
  warehouseId: number,
  params: WarehouseMaterialSearch,
): Promise<PageResult<WarehouseMaterial>> {
  return request.get(`/api/v1/warehouses/${warehouseId}/materials`, { params })
}

export function listAssignableMaterials(
  warehouseId: number,
  params: WarehouseMaterialSearch,
): Promise<PageResult<WarehouseMaterial>> {
  return request.get(`/api/v1/warehouses/${warehouseId}/assignable-materials`, { params })
}

export function assignMaterials(warehouseId: number, materialIds: number[]): Promise<{ assigned_count: number }> {
  return request.post(`/api/v1/warehouses/${warehouseId}/assign`, { material_ids: materialIds })
}

export function unassignMaterials(warehouseId: number, materialIds: number[]): Promise<{ unassigned_count: number }> {
  return request.post(`/api/v1/warehouses/${warehouseId}/unassign`, { material_ids: materialIds })
}

export const COURSE_SUBJECT_LABEL: Record<string, string> = {
  python: 'Python',
  java: 'Java',
  cpp: 'C/C++',
}

File: backend/app/schemas/warehouse.py
from datetime import datetime

from pydantic import BaseModel, Field


class WarehouseCreate(BaseModel):
    name: str = Field(min_length=1, max_length=64)
    description: str | None = None
    warehouse_kind: str = Field(default="file_type", pattern="^(file_type|course)$")
    course_subject: str | None = Field(default=None, pattern="^(python|java|cpp)$")
    material_type: str = Field(default="pdf", pattern="^(pdf|txt|md|pptx)$")
    icon: str = Field(default="📦", max_length=32)
    color: str = Field(default="#409eff", max_length=16)
    sort_order: int = Field(default=0, ge=0)


class WarehouseUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=64)
    description: str | None = None
    material_type: str | None = Field(default=None, pattern="^(pdf|txt|md|pptx)$")
    icon: str | None = Field(default=None, max_length=32)
    color: str | None = Field(default=None, max_length=16)
    sort_order: int | None = Field(default=None, ge=0)


class WarehouseOut(BaseModel):
    id: int
    name: str
    description: str | None
    warehouse_kind: str
    course_subject: str | None
    material_type: str
    icon: str
    color: str
    sort_order: int
    material_count: int = 0
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class WarehouseMaterialOut(BaseModel):
    id: int
    course_id: int
    course_name: str
    warehouse_id: int | None
    type: str
    original_name: str
    status: str
    uploaded_by: int | None
    uploader_name: str
    created_at: datetime

    model_config = {"from_attributes": True}


class WarehouseAssignBody(BaseModel):
    material_ids: list[int] = Field(min_length=1)
